feat: Add support for receiving gzip event payloads

keelerm84 opened the pull request:
BEGIN_COMMIT_OVERRIDE
feat: Support receiving compressed event payloads (#364)
feat: Add new `MaxInboundPayloadSize` configuration to limit event payload sizes (#364)
feat: gzip event payloads before sending to upstream LaunchDarkly APIs (#364)
fix: Bump minimum go version to 1.21 (#364)
END_COMMIT_OVERRIDE

keelerm84 committed:
feat: Add support for receiving gzip event payloads

keelerm84 commented on config/config.go:
```diff
@@ -3,6 +3,7 @@ package config
 import (
 	"time"
 
+	"github.com/alecthomas/units"
```
Adding this library because it allows for a nice size type (e.g. 15GiB).

keelerm84 commented on config/config.go:
```diff
@@ -179,11 +180,12 @@ type OfflineModeConfig struct {
+	FlushInterval            ct.OptDuration           `conf:"EVENTS_FLUSH_INTERVAL"`
+	Capacity                 ct.OptIntGreaterThanZero `conf:"EVENTS_CAPACITY"`
+	InlineUsers              bool                     `conf:"EVENTS_INLINE_USERS"`
+	MaxReceivablePayloadSize units.Base2Bytes         `conf:"EVENTS_MAX_RECEIVABLE_PAYLOAD_SIZE"`
```
I would welcome a better name.

cwaldren-ld replied:
How about: `MaxInboundPayloadSize` and `EVENTS_MAX_INBOUND_PAYLOAD_SIZE`?


If we added a config for the max that could be sent to LD, we could call it `Outbound`.

keelerm84 replied:
Updated to this new name.

keelerm84 commented on internal/util/reader.go:
```diff
@@ -0,0 +1,107 @@
+func NewReader(r io.ReadCloser, isGzipped bool, maxBytes int64) (io.ReadCloser, error) {
+	// If this isn't compressed, and we don't want to limit the size, just
+	// return the original reader
+	if !isGzipped && maxBytes == 0 {
```
To maintain backwards compatibility, if the new max payload size isn't specified,  it defaults to 0 which we will interpret to mean no restrictions on the size of the payload (the current behavior).

Because of this, we have a couple of early return chances here -- line 30 and 48. Because we are returning different types, the return type on line 26 is now an `io.ReadCloser` instead of the `PayloadReader` like one might expect.

This isn't exactly optimal. In a future release, when we don't mind breaking backwards compatibility, we won't have to deal with this 0 nonsense, and the logic can basically fall back through.

Alternatively, what do you think about making the config optional? If it isn't set, then we add no restriction, but once you specify a value, we enforce that limit, including 0.

kinyoklion replied:
We have some existing payload size in the megabytes that will cause the service to send a 413. It seems like we could reasonably have a default limit that is some size larger than that.

cwaldren-ld replied:
I like the idea that to maintain backwards compat, we allow the max payload size to not be set. 

If not set, then there is no restriction: old behavior.

If it is set, then it is restricted to whatever they choose.  We _could_ use -1 or 0 as a sentinel value to indicate "no restriction", but I think it would probably be better to enforce a positive number. That way you can be guaranteed that if you've configured this thing, you are protected to some extent.

cwaldren-ld replied:
Then in a major version, we start setting it automatically to some [well defined but suitable] max value (whatever ryan mentioned?).

keelerm84 replied:
Update the code to use the new configtypes.OptBase2Bytes type. This lets it default to nothing being enforced, but once it is set, it must be > 0 so we can do some simple enforcing.

keelerm84 commented on internal/util/reader_test.go:
```diff
@@ -0,0 +1,77 @@
+	nonZipBytes := []byte("this is a test")
+
+	reader, _ := NewReader(io.NopCloser(bytes.NewReader(nonZipBytes)), false, 1000)
+	payloadReader, _ := reader.(*PayloadReader)
```
Since we aren't returning the PayloadReader type, we have to do this typehint stuff in a couple of places here.

kinyoklion commented on internal/util/reader_test.go:
```diff
@@ -0,0 +1,77 @@
+
+func TestPayloadBytesTracking(t *testing.T) {
+	// build a string
+	var ret string
+	for i := 0; i < 500; i++ {
+		ret += "00"
+	}
```
`ret := strings.Repeat("0", 1000)`

keelerm84 committed:
Rename

keelerm84 committed:
tighten up tests

keelerm84 committed:
Replace with OptBase2Bytes type

This allows for better backwards compatibility in that the default
option is to not enforce a limit (current behavior). But once a customer
does set the value, we can enforce that as a strict > 0 limit.

keelerm84 committed:
Use published configtypes release

keelerm84 committed:
Go 1.20 to 1.21

keelerm84 committed:
rename variables

cwaldren-ld commented on internal/util/reader_test.go:
```diff
@@ -0,0 +1,73 @@
+}
+
+func TestZipBombing(t *testing.T) {
+	// build a string
```
Probably don't need this comment

keelerm84 committed:
remove comment

cwaldren-ld commented on go.mod:
```diff
@@ -1,6 +1,6 @@
 module github.com/launchdarkly/ld-relay/v8
 
-go 1.20
+go 1.21
```
Checking that this change actually requires a bump to the minimum go version?

github-actions[bot] committed:
chore(v8): release 8.8.1 (#414)

:robot: I have created a release *beep* *boop*
---


##
[8.8.1](https://github.com/launchdarkly/ld-relay/compare/v8.8.0...v8.8.1)
(2024-07-10)


### Bug Fixes

* **deps:** bump supported Go versions to 1.22.5 and 1.21.12
([#411](https://github.com/launchdarkly/ld-relay/issues/411))
([02c0a7e](https://github.com/launchdarkly/ld-relay/commit/02c0a7eb93ece5c288aaa754f9ec08785d0631e6))

---
This PR was generated with [Release
Please](https://github.com/googleapis/release-please). See
[documentation](https://github.com/googleapis/release-please#release-please).

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

sheenacarswell committed:
fix: redact password in logs if specified as part of URL (#413)

**Requirements**

- [x] I have added test coverage for new or changed functionality
- [x] I have followed the repository's [pull request submission
guidelines](../blob/master/CONTRIBUTING.md#submitting-pull-requests)
- [x] I have validated my changes against all supported platform
versions

**Related issues**

(Internal: see sc-249267)

**Describe the solution you've provided**

When the connection URL is specified with a username and password, this
was logged in full, for example "Using proxy server at
http://my-user-name:my-password@my-proxy-server". This fix redacts any
password specified as part of the URL from the logs.

github-actions[bot] committed:
chore(v8): release 8.8.2 (#416)

:robot: I have created a release *beep* *boop*
---


##
[8.8.2](https://github.com/launchdarkly/ld-relay/compare/v8.8.1...v8.8.2)
(2024-07-15)


### Bug Fixes

* redact password in logs if specified as part of URL
([#413](https://github.com/launchdarkly/ld-relay/issues/413))
([0471d51](https://github.com/launchdarkly/ld-relay/commit/0471d51a315637df91d38f42afedd0e170cdefad))

---
This PR was generated with [Release
Please](https://github.com/googleapis/release-please). See
[documentation](https://github.com/googleapis/release-please#release-please).

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

keelerm84 committed:
go v1.20

keelerm84 committed:
enable gzip everywhere

keelerm84 committed:
Bump go-sdk-events to 3.4.0 and fix tests

keelerm84 committed:
bump sdk version to 7.6.0

keelerm84 committed:
Merge branch 'v8' into mk/sc-244987/read-limit
